crates/shared/src/recent_block_cache.rs

```diff
@@ -180,33 +184,67 @@ where
             .keys_of_recently_used_entries()
             .collect::<HashSet<_>>();
         tracing::debug!("automatically updating {} entries", keys.len());
-        let entries = self
-            .fetch_inner(keys.clone(), Block::Number(new_block))
+        let (found_values, _) = self
+            .fetch_inner_many(keys.clone(), Block::Number(new_block))
             .await?;
-        {
-            let mut mutexed = self.mutexed.lock().unwrap();
-            mutexed.insert(new_block, keys.into_iter(), entries);
-            let oldest_to_keep = new_block.saturating_sub(self.number_of_blocks_to_cache.get() - 1);
-            mutexed.remove_cached_blocks_older_than(oldest_to_keep);
-            mutexed.last_update_block = new_block;
-        }
+
+        let mut mutexed = self.mutexed.lock().unwrap();
+        mutexed.insert(new_block, keys.into_iter(), found_values);
+        let oldest_to_keep = new_block.saturating_sub(self.number_of_blocks_to_cache.get() - 1);
+        mutexed.remove_cached_blocks_older_than(oldest_to_keep);
+        mutexed.last_update_block = new_block;
+
         Ok(())
     }
 
+    async fn fetch_inner_many(&self, keys: HashSet<K>, block: Block) -> Result<(Vec<V>, Vec<K>)> {
+        let fetched =
+            futures::future::join_all(keys.iter().map(|key| self.fetch_inner(key.clone(), block)))
+                .await;
+        let found_keys: Vec<_> = fetched
+            .iter()
+            .zip(keys.iter())
+            .filter_map(|(results, key)| {
+                results
+                    .as_ref()
+                    .is_ok_and(|res| !res.is_empty())
+                    .then_some(key.clone())
+            })
+            .collect();
+        let fetched: Vec<_> = fetched
+            .into_iter()
+            .filter_map(|res| res.ok())
+            .flatten()
+            .collect();
+        Ok((fetched, found_keys))
+    }
+
     // Sometimes nodes requests error when we try to get state from what we think is
     // the current block when the node has been load balanced out to one that
     // hasn't seen the block yet. As a workaround we repeat the request up to N
     // times while sleeping in between.
-    async fn fetch_inner(&self, keys: HashSet<K>, block: Block) -> Result<Vec<V>> {
-        let fetch = || self.fetcher.fetch_values(keys.clone(), block);
-        for _ in 0..self.maximum_retries {
-            match fetch().await {
-                Ok(values) => return Ok(values),
-                Err(err) => tracing::warn!("retrying fetch because error: {:?}", err),
+    async fn fetch_inner(&self, key: K, block: Block) -> Result<Vec<V>> {
+        let retries = self.maximum_retries;
+        let delay = self.delay_between_retries;
+        let fetcher = self.fetcher.clone();
+        let mut created = false;
```

Comment:
doesn't seem to be read anywhere.

Reply:
Right. That's left over debugging logic.